MISRA doc for CSDK hub repository. (#1238)
aws/aws-iot-device-sdk-embedded-C: MISRA.md

| Directive 4.9 | Advisory | Allow inclusion of function like macros. The `assert` macro is used throughout the library for parameter validation, and logging is done using function like macros. |
| Rule 2.3 | Advisory | Allow unused types. The `MQTTSubAckStatus_t` enum is unused in our source files, as it is intended for a user to use when parsing a subscription acknowledgment's response codes. |
| Rule 2.4 | Advisory | Allow unused tags. Some compilers warn if types are not tagged. |
| Rule 2.5 | Advisory | Allow unused macros. Library headers may define macros intended for the application's use, but are not used by a specific file. |
| Rule 3.1 | Required | Allow nested comments. C++ style `//` comments are used in example code within Doxygen documentation blocks. |
| Rule 11.5 | Advisory | Allow casts from `void *`. Fields such as publish payloads are passed as `void *` and must be cast to the correct data type before use. |
| Rule 21.1 | Required | Allow use of all macro names. For compatibility, some macros introduced in C99 are defined for use with C90 compilers. |

> | Rule 21.2 | Required | Allow use of all macro and identifier names. For compatibility, some macros introduced in C99 are defined for use with C90 compilers. |

Author: No. So the suggestion sounds good.

### Flagged by Coverity
| Deviation | Category | Justification |
| :-: | :-: | :-: |
| Rule 5.7 | Required | The network context struct is defined by each transport implementation; however, only one transport implementation is needed by an application at a time. If the transport implementations are analyzed one at a time by Coverity, this violation will not be flagged. |
| Rule 8.7 | Advisory | API functions are not used by the library outside of the files they are defined; however, they must be externally visible in order to be used by an application. |

### Suppressed with Coverity Comments
| Deviation | Category | Justification |
| :-: | :-: | :-: |
| Directive 4.6 | Advisory | Basic numerical types are used in platform implementations, since the POSIX platform functions and openssl functions used in the platform implementations take arguments of these types. |

> | Rule 10.1 | Required | A POSIX specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementations is supplied by the system, is used in the transport implementation. |
> | Rule 10.8 | Required | A POSIX specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementations is supplied by the system, is used in the transport implementation. |

Reviewer: ```suggestion
| Rule 10.1 | Required | A POSIX-specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementations are supplied by the system, is used in the transport implementation. |
| Rule 10.8 | Required | A POSIX-specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementations are supplied by the system, is used in the transport implementation. |
```

Author: Same comment.

> | Rule 11.3 | Required | The transport implementation casts an object pointer to a pointer of a different object type. This cast is supported in POSIX, and is used to obtain the IP address from an address record. |

Reviewer: I see you changed `an` back to `the`. I didn't like `the` since it implies we've referenced it before. Here's another suggestion:
```suggestion
| Rule 11.3 | Required | The transport implementation casts an object pointer to a pointer of a different object type. This cast is supported in POSIX, and is used to obtain IP addresses from address records. |
```

Author: I didn't like the 'an' there. The new suggestion is good. Thanks.

> | Rule 11.3 | Required | Transport implementation casts a pointer of a object type to a pointer of a different object type. This casting is supported in POSIX and is used to obtain the IP address from a address record. |

Reviewer: ```suggestion
| Rule 11.3 | Required | The transport implementation casts an object pointer to a pointer of a different object type. This cast is supported in POSIX, and is used to obtain an IP address from an address record. |
```

> | Rule 11.8 | Required | An openssl API `SSL_set_tlsext_host_name`, which is used in TLS transport implementation internally casts a string literal to a `void *` pointer. |

Reviewer: ```suggestion
| Rule 11.8 | Required | An openssl API `SSL_set_tlsext_host_name`, which is used in the TLS transport implementation, internally casts a string literal to a `void *` pointer. |
```

> | Rule 11.8 | Required | An openssl API used in TLS transport implementation internally casts the pointer to a string literal to a `void *` pointer. |

Reviewer: ```suggestion
| Rule 11.8 | Required | An openssl API used in TLS transport implementation internally casts a string literal to a `void *` pointer. |
```

> | Rule 10.1 | Required | A POSIX specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |
> | Rule 10.8 | Required | A POSIX specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |
> | Rule 11.3 | Required | The transport implementation casts an object pointer to a pointer of a different object type. This cast is supported in POSIX, and is used to obtain IP addresses from address records. |
> | Rule 11.8 | Required | An openssl API `SSL_set_tlsext_host_name`, which is used in the TLS transport implementation, internally casts a string literal to a `void *` pointer. |
> | Rule 13.4 | Required | A POSIX specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |
> | Rule 14.4 | Required | A POSIX specific macro utility `FD_ZERO` is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |

Reviewer: I'd prefer a hyphen before "specific", but this seems like such a minor thing to nitpick on
```suggestion
| Rule 10.1 | Required | A POSIX-specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |
| Rule 10.8 | Required | A POSIX-specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |
| Rule 11.3 | Required | The transport implementation casts an object pointer to a pointer of a different object type. This cast is supported in POSIX, and is used to obtain IP addresses from address records. |
| Rule 11.8 | Required | An openssl API `SSL_set_tlsext_host_name`, which is used in the TLS transport implementation, internally casts a string literal to a `void *` pointer. |
| Rule 13.4 | Required | A POSIX-specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |
| Rule 14.4 | Required | A POSIX-specific macro utility `FD_ZERO` is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |
```

> | Rule 13.4 | Required | A POSIX specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementations is supplied by the system, is used in the transport implementation. |
> | Rule 14.4 | Required | A POSIX specific macro utility `FD_ZERO` is flagged for this violation. This macro utility, whose implementations is supplied by the system, is used in the transport implementation. |

Reviewer: ```suggestion
| Rule 13.4 | Required | A POSIX-specific macro utility `FD_SET` is flagged for this violation. This macro utility, whose implementations are supplied by the system, is used in the transport implementation. |
| Rule 14.4 | Required | A POSIX-specific macro utility `FD_ZERO` is flagged for this violation. This macro utility, whose implementations are supplied by the system, is used in the transport implementation. |
```

Author: Thanks for catching this. I think `whose implementation is` is correct. So fixing it locally.

> | Rule 14.4 | Required | POSIX specific macro utility is flagged for this violation. This macro utility, whose implementation is supplied by the system, is used in the transport implementation. |

Reviewer: I see a lot of `POSIX specific macro utility`. Do we want to specify which one? If it's multiple, I'd make it plural. If not, it at least should have an indefinite article ('a')

Author: It is multiple. I will update the sentence. Thanks

> | Rule 21.6 | Required | The Standard Library input/output functions for opening and closing a file are used by the openssl transport implementation, as the openssl API to read a PEM file takes `FILE *` as an argument. |

Reviewer: ```suggestion
| Rule 21.6 | Required | The Standard Library input/output functions for opening and closing files are used by the openssl transport implementation, since the openssl API to read PEM files takes `FILE *` as an argument. |
```
